File: patches/fix_tray_dbus.py
#!/usr/bin/env python3
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: python
"""
Patch Claude Desktop tray menu handler to prevent DBus race conditions.

The tray icon setup can be called multiple times concurrently, causing DBus
"already exported" errors. This patch:
1. Makes the tray function async
2. Adds a mutex guard to prevent concurrent calls
3. Adds a delay after Tray.destroy() to allow DBus cleanup

Based on: https://github.com/aaddrick/claude-desktop-debian/blob/main/build.sh

Usage: python3 fix_tray_dbus.py <path_to_index.js>
"""

import sys
import os
import re


def patch_tray_dbus(filepath):
    """Patch the tray menu handler to prevent DBus race conditions."""

    print(f"Patching tray DBus handler in: {filepath}")

    with open(filepath, 'rb') as f:
        content = f.read()

    original_content = content
    patches_applied = 0

    # Step 1: Find the tray function name from menuBarEnabled listener
    # Pattern: on("menuBarEnabled",()=>{FUNCNAME()})
    match = re.search(rb'on\("menuBarEnabled",\(\)=>\{(\w+)\(\)\}\)', content)
    if not match:
        print("  Warning: Could not find menuBarEnabled listener")
        return False

    tray_func = match.group(1)
    print(f"  Found tray function: {tray_func.decode()}")

    # Step 2: Find tray variable name
    # Pattern: });let TRAYVAR=null;function FUNCNAME or });let TRAYVAR=null;async function FUNCNAME
    pattern = rb'\}\);let (\w+)=null;(?:async )?function ' + tray_func
    match = re.search(pattern, content)
    if not match:
        print("  Warning: Could not find tray variable")
        return False

    tray_var = match.group(1)
    print(f"  Found tray variable: {tray_var.decode()}")

    # Step 3: Make the function async (if not already)
    old_func = b'function ' + tray_func + b'(){'
    new_func = b'async function ' + tray_func + b'(){'
    if old_func in content and b'async function ' + tray_func not in content:
        content = content.replace(old_func, new_func)
        patches_applied += 1
        print(f"  Made {tray_func.decode()}() async")

    # Step 4: Find first const variable in the function
    # Pattern: async function FUNCNAME(){const VARNAME= or with mutex already
    pattern = rb'async function ' + tray_func + rb'\(\)\{(?:if\(' + tray_func + rb'\._running\)[^}]*?)?const (\w+)='
    match = re.search(pattern, content)
    if not match:
        print("  Warning: Could not find first const in function")
        return False

    first_const = match.group(1)
    print(f"  Found first const: {first_const.decode()}")

    # Step 5: Add mutex guard (if not already present)
    mutex_check = tray_func + b'._running'
    if mutex_check not in content:
        old_start = b'async function ' + tray_func + b'(){const ' + first_const + b'='
        mutex_code = (
            b'async function ' + tray_func + b'(){if(' + tray_func + b'._running)return;' +
            tray_func + b'._running=true;setTimeout(()=>' + tray_func + b'._running=false,500);const ' +
            first_const + b'='
        )
        if old_start in content:
            content = content.replace(old_start, mutex_code)
            patches_applied += 1
            print(f"  Added mutex guard to {tray_func.decode()}()")
    else:
        print(f"  Mutex guard already present")

    # Step 6: Add delay after Tray.destroy() for DBus cleanup
    # Pattern: TRAYVAR&&(TRAYVAR.destroy(),TRAYVAR=null)
    old_destroy = tray_var + b'&&(' + tray_var + b'.destroy(),' + tray_var + b'=null)'
    new_destroy = tray_var + b'&&(' + tray_var + b'.destroy(),' + tray_var + b'=null,await new Promise(r=>setTimeout(r,50)))'

    if old_destroy in content and b'await new Promise' not in content:
        content = content.replace(old_destroy, new_destroy)
        patches_applied += 1
        print(f"  Added DBus cleanup delay after {tray_var.decode()}.destroy()")
    elif b'await new Promise' in content:
        print(f"  DBus cleanup delay already present")

    if content != original_content:
        with open(filepath, 'wb') as f:
            f.write(content)
        print(f"Tray DBus patches applied: {patches_applied}")
        return True
    else:
        print("Warning: No tray DBus patches applied")
        return False


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <path_to_index.js>")
        sys.exit(1)

    filepath = sys.argv[1]
    if not os.path.exists(filepath):
        print(f"Error: File not found: {filepath}")
        sys.exit(1)

    patch_tray_dbus(filepath)
    # Always exit 0 - patch is best-effort
    sys.exit(0)

File: patches/fix_tray_dbus.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
# Patch tray menu handler to prevent DBus race conditions on Linux.
# Makes tray function async, adds mutex guard, adds delay after destroy.

import std/[os, strformat, strutils]
import regex

proc apply*(input: string): string =
  result = input
  var failed = false

  # Step 1: Find the tray function name from menuBarEnabled listener
  let listenerPat = re2"on\(""menuBarEnabled"",\(\)=>\{(\w+)\(\)\}\)"
  var trayFunc = ""
  for m in result.findAll(listenerPat):
    trayFunc = result[m.group(0)]
    break

  if trayFunc == "":
    echo "  [FAIL] menuBarEnabled listener: 0 matches, expected >= 1"
    failed = true
  else:
    echo &"  [OK] menuBarEnabled listener: found tray function '{trayFunc}'"

  # Step 2: Find tray variable name
  var trayVar = ""
  if trayFunc != "":
    let varPat = re2("let ([\\w$]+)=null;(?:async )?function " & trayFunc)
    for m in result.findAll(varPat):
      trayVar = result[m.group(0)]
      break

    if trayVar == "":
      echo "  [FAIL] tray variable: 0 matches, expected >= 1"
      failed = true
    else:
      echo &"  [OK] tray variable: found '{trayVar}'"

  # Step 3: Make the function async (if not already)
  if trayFunc != "":
    let oldFunc = "function " & trayFunc & "(){"
    let newFunc = "async function " & trayFunc & "(){"
    let asyncCheck = "async function " & trayFunc & "(){"
    if oldFunc in result and asyncCheck notin result:
      result = result.replace(oldFunc, newFunc)
      echo &"  [OK] async conversion: made {trayFunc}() async"
    elif asyncCheck in result:
      echo "  [INFO] async conversion: already async"
    else:
      echo "  [FAIL] async conversion: function pattern not found"
      failed = true

  # Step 4: Find first const variable in the function
  var firstConst = ""
  if trayFunc != "":
    let constPat = re2("async function " & trayFunc & "\\(\\)\\{.+?const (\\w+)=")
    for m in result.findAll(constPat):
      firstConst = result[m.group(0)]
      break

    if firstConst == "":
      echo "  [FAIL] first const in function: 0 matches"
      failed = true
    else:
      echo &"  [OK] first const in function: found '{firstConst}'"

  # Step 5: Add mutex guard
  if trayFunc != "" and firstConst != "":
    let mutexCheck = trayFunc & "._running"
    if mutexCheck notin result:
      let oldStart = "async function " & trayFunc & "(){"
      let mutexPrefix = "async function " & trayFunc & "(){if(" & trayFunc & "._running)return;" & trayFunc & "._running=true;setTimeout(()=>" & trayFunc & "._running=false,500);"
      if oldStart in result:
        result = result.replace(oldStart, mutexPrefix)
        echo "  [OK] mutex guard: added"
      else:
        echo "  [FAIL] mutex guard: insertion point not found"
        failed = true
    else:
      echo "  [INFO] mutex guard: already present"

  # Step 6: Add delay after Tray.destroy() for DBus cleanup
  if trayVar != "":
    let oldDestroy = trayVar & "&&(" & trayVar & ".destroy()," & trayVar & "=null)"
    let newDestroy = trayVar & "&&(" & trayVar & ".destroy()," & trayVar & "=null,await new Promise(r=>setTimeout(r,50)))"

    if oldDestroy in result and newDestroy notin result:
      result = result.replace(oldDestroy, newDestroy)
      echo &"  [OK] DBus cleanup delay: added after {trayVar}.destroy()"
    elif newDestroy in result:
      echo "  [INFO] DBus cleanup delay: already present"
    else:
      echo "  [FAIL] DBus cleanup delay: destroy pattern not found"
      failed = true

  if failed:
    raise newException(ValueError, "fix_tray_dbus: Some required patterns did not match")

when isMainModule:
  if paramCount() != 1:
    echo "Usage: fix_tray_dbus <file>"
    quit(1)
  let file = paramStr(1)
  echo "=== Patch: fix_tray_dbus ==="
  echo &"  Target: {file}"
  if not fileExists(file):
    echo &"  [FAIL] File not found: {file}"
    quit(1)
  let input = readFile(file)
  let output = apply(input)
  if output != input:
    writeFile(file, output)
    echo "  [PASS] All required patterns matched and applied"
  else:
    echo "  [WARN] No changes made (patterns may have already been applied)"
